sdk/digitaltwins/digitaltwins/src/digitalTwinsClient.ts

```diff
@@ -0,0 +1,851 @@
+  EventRoutesListNextResponse,
+  EventRoute,
+  EventRoutesAddOptionalParams,
+  QueryQueryTwinsResponse,
+  QuerySpecification
+} from "./generated/src/models";
+
+/**
+ * Client for Azure IoT DigitalTwins API.
+ */
+export class DigitalTwinsClient {
+  /**
+   * @internal
+   * @ignore
+   * A reference to the auto-generated AzureDigitalTwinsAPI
+   */
+  private readonly client: GeneratedClient;
+
+  // private readonly client!: AzureDigitalTwinsAPI;
+
+  /**
+   * Creates an instance of AzureDigitalTwinsAPI.
+   *
+   * Example usage:
+   * ```ts
+   * const { DigitalTwinsClient, ServiceClientCredentials } = require("@azure/digitaltwins");
+   *
+   * const client = new DigitalTwinsClient(
+   *   "<endpoint>",
+   *   new DefaultAzureCredential();
+   * );
+   * ```
+   * @param {string} endpoint The endpoint of the service. If options parameter is given than this value will be ignored.
+   * @param {TokenCredential | ServiceClientCredentials} credential Used to authenticate requests to the service.
+   * @param {AzureDigitalTwinsAPIOptions} [options] Used to configure the service client.
+   */
+  constructor(
+    enpoint: string,
+    credential: TokenCredential | ServiceClientCredentials,
+    options?: AzureDigitalTwinsAPIOptions
+  ) {
+    if (!options) {
+      const options: AzureDigitalTwinsAPIOptions = {};
+      options.baseUri = enpoint;
+    }
+
+    this.client = new GeneratedClient(credential, options);
+  }
+
+  /**
+   * Get a digital twin
+   *
+   * @param digitalTwinId The Id of the digital twin.
+   * @returns The application/json digital twin and the http response.
+   */
+  public getDigitalTwin(digitalTwinId: string): Promise<DigitalTwinsGetByIdResponse> {
+    return this.client.digitalTwins.getById(digitalTwinId);
+  }
+
+  /**
+   * Create or update a digital twin
+   *
+   * @param digitalTwinId The Id of the digital twin to create or update.
+   * @param digitalTwin The application/json digital twin to create.
+   * @param enableUpdate If true then update of an existing digital twin is enabled.
+   * @returns The created application/json digital twin and the http response.
+   */
+  public upsertDigitalTwin(
+    digitalTwinId: string,
+    digitalTwin: string,
+    enableUpdate: boolean = true
+  ): Promise<DigitalTwinsAddResponse> {
+    var options = <DigitalTwinsAddOptionalParams>{};
+    if (!enableUpdate) {
+      options.ifNoneMatch = "*";
+    }
+    return this.client.digitalTwins.add(digitalTwinId, digitalTwin, options);
+  }
+
+  /**
+   * Update a digital twin using a json patch.
+   *
+   * @param digitalTwinId The Id of the digital twin.
+   * @param twinPatch An update specification described by JSON Patch. Updates to property values
+   * and $model elements may happen in the same request. Operations are limited to add, replace and
+   * remove.
+   * @returns The http response.
+   */
+  public updateDigitalTwin(
+    digitalTwinId: string,
+    twinPatch: any,
```

Comment:
not sure how good they are, but seems like there are some definitions out there for patch: https://www.npmjs.com/package/@types/json-patch

Reply:
The type of the patch is coming from the generated code. We had discussion with the service team and they want to keep it "any"

---
In reply to: [447261717](https://github.com/Azure/azure-sdk-for-js/pull/9787#discussion_r447261717) [](ancestors = 447261717)